### PR TITLE
Fix counter metrics for ParDo#with_exception_handling(timeout).

### PR DESCRIPTION
**Pull request by claudevdm**
Make parent thread state sampler available in child thread when running ParDo#with_exception_handling(timeout).

- ParDo#with_exception_handling(timeout) executes the DoFn in a new thread [1]
- Incrimenting a counter invokes tries to access the state sampler that was declared in the main thread local storage from the child thread [2]
- Since the child thread does not have access to the state sampler from the main thread the get_current_tracker method fails silently [3] and the counter isnt actually incremented

This change makes the parent thread's state sampler available to the child thread.

[1] https://github.com/apache/beam/blob/1eddbdca4c5efe4bfccc11af29624f9689ea2997/sdks/python/apache_beam/transforms/core.py#L2599
[2] https://github.com/apache/beam/blob/1eddbdca4c5efe4bfccc11af29624f9689ea2997/sdks/python/apache_beam/runners/worker/statesampler.py#L43
[3] https://github.com/apache/beam/blob/1eddbdca4c5efe4bfccc11af29624f9689ea2997/sdks/python/apache_beam/runners/worker/statesampler.py#L54 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: sdks/python/apache_beam/transforms/core.py [hunk 1]
```diff
@@ -2594,11 +2594,23 @@ def __getattribute__(self, name):
...
       self._pool = concurrent.futures.ThreadPoolExecutor(10)
+
+    # Import here to avoid circular dependency
+    from apache_beam.runners.worker.statesampler import get_current_tracker, set_current_tracker
+
+    # State sampler/tracker is stored as a thread local variable, and is used
+    # when incrementing counter metrics.
+    dispatching_thread_state_sampler = get_current_tracker()
+
+    def wrapped_process():
```
Comment (tvalentyn) [inline sdks/python/apache_beam/transforms/core.py:2605]: Have you considered an idea where wrapped_process would be returning a tuple:

     (_fn_process_output_list, modifications_to_global_state_that_can_be_subsequently_reapplied_onto_the_parent_thread) ? 

There are questions how to capture the modifications of global state, and what what the exact modifications we will be watching for. 

Something that comes to mind is a possibility of temporarily using stubs for objects/methods. For example unittest.mock often employs this technique to stub-out portions of the codebase that are not under test. In our scenario, we could consider  setting stubs for elements of the global state, and stubs would instead collect the necessary updates, that we can reapply on the main thread or process after the subprocess call finishes. 

In this scenario, we might be able to pass the information also across the process boundary, not just the thread boundary.

File: sdks/python/apache_beam/transforms/core.py [hunk 2]
```diff
@@ -2594,11 +2594,23 @@ def __getattribute__(self, name):
   def process(self, *args, **kwargs):
     if self._pool is None:
       self._pool = concurrent.futures.ThreadPoolExecutor(10)
+
+    # Import here to avoid circular dependency
+    from apache_beam.runners.worker.statesampler import get_current_tracker, set_current_tracker
+
+    # State sampler/tracker is stored as a thread local variable, and is used
+    # when incrementing counter metrics.
+    dispatching_thread_state_sampler = get_current_tracker()
```
Comment (tvalentyn) [inline sdks/python/apache_beam/transforms/core.py:2603]: can this code be in `__init__` instead of `process()` ?

Comment (claudevdm) [reply sdks/python/apache_beam/transforms/core.py:2620]: The current tracker is not set yet when the _TimeoutDoFn is constructed so adding this to __init__ results in a None tracker.

File: sdks/python/apache_beam/transforms/core.py (hunk 1 above)
Comment (claudevdm) [reply sdks/python/apache_beam/transforms/core.py:2622]: Implemented this solution in https://github.com/apache/beam/pull/32600, but as discussed offline we will keep the wrapper approach for the threading case and try the stub approach for the subprocess case.

**Review by tvalentyn — APPROVED** [review]

**Pull request merged**